src/functions_math.jl

```diff
@@ -108,3 +113,131 @@ function symmetric_names(L::Tuple{Symbol,Symbol}, dims::Integer)
+        end
+
+        function AbstractFFTs.$fun(A::NamedDimsArray{L,T,N}, dims) where {L,T,N}
+            numerical_dims = dim(A, dims)
+            data = AbstractFFTs.$fun(parent(A), numerical_dims)
+            newL = wave_name(L, numerical_dims)
+            return NamedDimsArray(data, newL)
+        end
+
+        """
+            $($fun)(A, :time => :freq, :x => :kx)
+
+        Acting on a `NamedDimsArray`, this specifies to take the transform along the dimensions
+        named `:time, :x`, and return an array with names `:freq` and `:kx` in their places.
+
+        If new names are not given, then the default is `:x => :x∿` and `:x∿ => :x`,
+        applied to all dimensions, or to those specified as usual, e.g. `$($fun)(A, (1,2))`
+        or `$($fun)(A, :time)`. The symbol "∿" can be typed by `\\sinewave<tab>`.
+        """
+        function AbstractFFTs.$fun(A::NamedDimsArray{L,T,N}, p::Pair{Symbol,Symbol}, ps::Pair{Symbol,Symbol}...) where {L,T,N}
+            numerical_dims = dim(A, (first(p), first.(ps)...))
+            data = AbstractFFTs.$fun(parent(A), numerical_dims)
+            newL = replace_names(L, p, ps...)
+            return NamedDimsArray(data, newL)
+        end
+
+        """
+            F = $($fun)(A, :name)
+            A∿ = F * A == $($fun)(A, :name)
+
+        FFT plans for `NamedDimsArray`s can select a dimension based on names,
+        but cannot specify the final name. The `A∿` returned here will have default name
+        `:name∿` in the transformed dimension.
+        """
+        function AbstractFFTs.$plan_fun(A::NamedDimsArray, dims = ntuple(d->d, ndims(A)); kw...)
+            numerical_dims = dim(A, dims)
+            return AbstractFFTs.$plan_fun(parent(A), numerical_dims; kw...)
+        end
+
+        AbstractFFTs.$plan_fun(A::NamedDimsArray, p::Pair{Symbol,Symbol}, ps::Pair{Symbol,Symbol}...) =
+            throw(ArgumentError(string($plan_fun, "(::NamedDimsArray, ::Pair...) is not implemented. ",
+                "FFT plans can only specify an input name, not an output.")))
+
+    end
+
+end
+
+for plan_type in (:Plan, :ScaledPlan)
+
+    @eval function Base.:*(plan::AbstractFFTs.$plan_type, A::NamedDimsArray{L,T,N}) where {L,T,N}
+        data = plan * parent(A)
+        if _hasproperty(plan, :region) # true for plan_fft from FFTW
+            dims = plan.region         # dims can be 1, (1,3) or 1:3
+        elseif _hasproperty(plan, :p)
+            dims = plan.p.region
+        else
+            return data
+        end
+        newL = ntuple(d -> d in dims ? wave_name(L[d]) : L[d], N)::NTuple{N,Symbol}
+        # plan.region is not part of the type, so using compile_time_return_hack is much slower:
+        # newL = wave_name(L, Tuple(dims)) # 37μs instead of 7.
+        return NamedDimsArray(data, newL)
+    end
+
+end
+
+if VERSION > v"1.1"
+    _hasproperty(x, s)::Bool = hasproperty(x, s)
+else
+    _hasproperty(x, s)::Bool = Base.sym_in(s, propertynames(x))
+end
```

Comment:
this should move to `src/compat.jl`
Or possibly we can just make use of the **Compat.jl** package for it?

Reply:
I can just always use ` Base.sym_in` inline. (I moved it out to see if it mattered for speed.)

The treatment of `plan_fft` seems a bit awkward to me. The alternative it to return a new struct `NamedPlan` or something, which would explicitly encode the names & dimensions, and make its application zero-cost again. But perhaps this adds more code than it deserves, thoughts?

Reply:
> I can just always use Base.sym_in inline. (I moved it out to see if it mattered for speed.)

That would be worse for readability.
just a 
```
if isdefined(Base, :hasproperty)
    hasproperty(x, s)::Bool = Base.sym_in(s, propertynames(x))
end
```

in the compat file.
(or check Compat.jl)

Reply:
I think `NamedPlan` make sense, but i don't know FFT well

Reply:
OK will tidy up the struct version then.

Reply:
Will get to this eventually, BTW. The struct version had a few unexpected complications.

Reply:
OK, I trashed the whole `NamedPlan` business, it got complicated. We can re-invent it if someone uses this & needs it, maybe make the basic case work first.